pkg/devicemapper/devmapper_log.go

```diff
@@ -1,21 +1,49 @@
-// +build linux
+// +build linux,cgo
 
 package devicemapper
 
 import "C"
 
 import (
+	"fmt"
 	"strings"
+
+	"github.com/sirupsen/logrus"
 )
 
+// DevmapperLogger defines methods required to register as a callback for
+// logging events recieved from devicemapper. Note that devicemapper will send
+// *all* logs regardless to callbacks (including debug logs) so it's
+// recommended to not spam the console with the outputs.
+type DevmapperLogger interface {
+	// DMLog is the logging callback containing all of the information from
+	// devicemapper. The interface is identical to the C libdm counterpart.
+	DMLog(level int, file string, line int, dmError int, message string)
+}
+
+// dmLogger is the current logger in use that is being forwarded our messages.
+var dmLogger DevmapperLogger
+
+// LogInit changes the logging callback called after processing libdm logs for
+// error message information. The default logger simply forwards all logs to
+// logrus. Calling LogInit(nil) disables the calling of callbacks.
+func LogInit(logger DevmapperLogger) {
+	dmLogger = logger
+}
+
 // Due to the way cgo works this has to be in a separate file, as devmapper.go has
 // definitions in the cgo block, which is incompatible with using "//export"
 
-// StorageDevmapperLogCallback exports the devmapper log callback for cgo.
-//export StorageDevmapperLogCallback
-func StorageDevmapperLogCallback(level C.int, file *C.char, line C.int, dmErrnoOrClass C.int, message *C.char) {
+// DevmapperLogCallback exports the devmapper log callback for cgo. Note that
+// because we are using callbacks, this function will be called for *every* log
+// in libdm (even debug ones because there's no way of setting the verbosity
+// level for an external logging callback).
+//export DevmapperLogCallback
+func DevmapperLogCallback(level C.int, file *C.char, line, dmErrnoOrClass C.int, message *C.char) {
```

Comment:
We need to leave this rename in place, to avoid symbol multiply-defined errors when this version of the library is pulled in by a newer containers/image into dockerd, per #14.

Reply:
Ok changed back

Reply:
No, it needs to be left renamed to `StorageDevmapperLogCallback`, as it was before this PR.